docs/phdata.rst

```diff
@@ -397,6 +407,60 @@ The following fields are also arrays:
   This field is not relevant when no polarization- and spectral-insensitive
   splitting is performed.
 
+The following fields can be used to indicate the meaning of space-time markers.
+
+- **num_space_time_markers**: (integer) *New in version 0.6.* number of
+  space-time markers. Required if space-time markers are recorded. This value
+  must match the number of space-time markers specified by
+  ``detectors_specs/space_time_marker1``,
+  ``detectors_specs/space_time_marker2``, and so on.
+
+- **space_time_markers**: (array of strings) *New in version 0.6.* for each
+  space-time marker, this field indicates a standard interpretation. The length
+  of this array must equal the value of ``num_space_time_markers``. Currently
+  defined interpretations are for indicating raster timings in laser-scanning
+  imaging (such as FLIM), and valid element values are:
+
+  - ``"pixel"`` (pixel clock; the marker indicates the beginning timestamp of
+    each raster pixel),
+  - ``"line"`` (line clock; the marker indicates the beginning timestamp of each
+    raster scanline),
+  - ``"frame"`` (frame clock; the marker indicates the beginning timestamp of
+    each raster frame), and
+  - ``""`` (empty string; indicates a marker not used for any of the above
+    standard meanings).
+
+.. _raster_group:
+
+Raster group
+^^^^^^^^^^^^
+
+*New in version 0.6.*
+
+The optional **/setup/raster** group contains information about raster scan
+measurements (typically FLIM, but could be any single-photon-counting
+raster-scan imaging method).
+
+These 3 fields are required if this group is present:
+
+- **raster_width**: (integer) width of raster scan (in pixels)
+- **raster_height**: (integer) height of raster scan (in pixels)
+- **pixel_time**: (float) the pixel period of the raster scan (in seconds)
```

Comment:
How necessary is this (and should it be required) as this information could theoretically be inferred from the marker timestamps, and could there be a FLIM imaging mode where different pixels have different times? I.e. maybe a setup where the acquisition software waits on a pixel until a certain number of photons are collected instead of a specific time to move onto the next pixel.

Reply:
Raster scans can be recorded with any of `(frame, line)`, `(frame, line, pixel)`, `(line, pixel)`, `(line,)`, or `(pixel,)` clocks/markers (and theoretically other combinations). I have seen most of these in actual use (not all scanners produce all 3 clocks, and on some TCSPC devices (Swabian) it helps to conserve marker channels). If you only have a line clock, for example, you need all 3 of these fields to recover an image. It is also possible to have a scanner that produces a constant pixel clock even during retrace, so that you need to count the pixels following each line marker.

Some of these could probably be made optional -- especially the `pixel_time` when a pixel clock is being recorded. I do think it is easier to record at least the width and height, though, because processing code will be more complex if it has to determine these as it goes (requires reallocating the image buffer or reading the data twice). (A minor point is that there is also the case where a partial frame is recorded at the end of the stream, and if the first frame happens to be partial, there is no way to determine how many lines there were supposed to be.)

I don't think waiting at a pixel for a photon count is very practical (with software in the loop, at least, due to high latency), but a resonant scanner is a case where a single `pixel_time` field won't work. My stance in writing this was that we don't (yet) support the case of resonant scanners. Maybe that could be added in the future as a `raster_type` other than `raster`.

Reply:
Interesting, I wasn't aware that different scans didn't record "all" markers.

In that case, then yes, it makes sense to require recording width/height/time markers.